docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+- `compiler_artifact.<artifactname>.outputdir = <dir>`
+
+When the second part of the key matches the name of a loaded generator, any artifacts produced by that generator will be placed in the specified `<dir>` instead of the default directory. If the key fails to match a generator, individual artifacts are checked: if the hint name of an artifact matches it will be placed in the specified `<dir>`.
+
+While these properties can be set manually in a `.globalconfig` it's expected that they will instead be passed from `MSBuild` as project items. The compiler already has a mechanism via `.MSBuildGeneratedEditorConfig` that allows the mapping of `MSBuild` information into `.globalconfig` format, and this proposal expands on that to support the compiler artifact mapping. The mapping will recognize the `<CompilerArtifactMapping` item, and will map the `include` value into the key name and the `Location` attribute as the key value (`<dir>`)
+
+```xml
+<ItemGroup>
+    <!-- Map all of the outputs of the razor source generator -->
+    <CompilerArtifactMapping Include="Microsoft.CodeAnalysis.Razor.SourceGenerators.SourceGenerator" Location="$(OutDir)\Razor" />
+
+    <!-- Map an individual artifact -->
+    <CompilerArtifactMapping Include="ArtifactName" Location="c:\dir" />
+<ItemGroup>
+```
+
+Would produce the following in the generated editor config:
+
+```ini
+compiler_artifact.Microsoft.CodeAnalysis.Razor.SourceGenerators.SourceGenerator.outputdir = bin\net8.0\Razor
+compiler_artifact.ArtifactName.outputdir = c:\dir
+```
+
+This means that generator authors are free to specify a mapping alongside their generator via the standard `NuGet` props and targets mechanism, and consumers of the generators are able to override the location in a consistent way.
+
+## MSBuild semantics
+
+Today, it's not possible to dynamically provide the list of outputs to an `MSBuild` target. This proposal allows the compiler to effectively emit an unbound number of files, none of which are tracked via `MSBuild` meaning that incremental compilation can be broken. (Note this is actually true today thanks to `EmitCompilerGeneratedFiles` but they don't have a bearing on the correctness of build).
+
+In order to correctly participate in the build process the compiler and targets are updated to emit a new file with the list of all files that it wrote as part of a compilation. A new target is added before `CoreCompile` that runs unconditionally and looks for the existence of the output list file; if the list is present and anything in the list is missing or out of date, the target touches a 'sentinel file'. The 'sentinel' file is passed as an input to `CoreCompile`. This ensures that anytime an output from the compiler is modified outside of the compilation process, `CoreCompile` will be correctly invoked and not skipped as being up to date. The 'sentinel' file effectively stands in for all of the outputs produced by the compiler in the `MSBuild` up to date checks.
```

Comment:
Think we should detail name of file and where it lives on disk. Assume it's in `obj` next to the outputted PE.